### PR TITLE
Don't allow to generate a method named with a PHP restricted keyword

### PR DESCRIPTION
**Pull request by mheki**
Fixes: https://github.com/phpspec/phpspec/issues/845

Instead of generating method with wrong name and failing it warns user about the problem: 
screen shot 2016-02-12 at 16 14 46

### CODE REVIEW

File: src/PhpSpec/Listener/MethodNotFoundListener.php [hunk 1]
```diff
@@ -27,12 +28,28 @@ class MethodNotFoundListener implements EventSubscriberInterface
...
+     * @param IO $io
+     * @param ResourceManagerInterface $resources
+     * @param GeneratorManager $generator
+     * @param VoterInterface $nameChecker
+     */
+    public function __construct(
+        IO $io,
+        ResourceManagerInterface $resources,
+        GeneratorManager $generator,
+        VoterInterface $nameChecker
```
Comment (ciaranmcnulty) [inline src/PhpSpec/Listener/MethodNotFoundListener.php]: Changing the constructor in this way is a BC break.

Please make this optional and instantiate a new copy inside the constructor if it's not provided (we can remove this in PhpSpec 3.0)

File: src/PhpSpec/Util/VoterInterface.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * (c) Marcello Duarte <marcello.duarte@gmail.com>
+ * (c) Konstantin Kudryashov <ever.zet@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace PhpSpec\Util;
+
+interface VoterInterface
```
Comment (ciaranmcnulty) [inline src/PhpSpec/Util/VoterInterface.php]: This naming is far too generic. Where you use it elsewhere it's a `nameChecker` which feels closer to a good name.

Find something that describes how this type is _used_, then for the concrete think about _how_ it does it

File: src/PhpSpec/Util/PhpNameChecker.php [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace PhpSpec\Util;
+
+final class PhpNameChecker implements VoterInterface
+{
+    private $reservedWords = array(
```
Comment (ciaranmcnulty) [inline src/PhpSpec/Util/PhpNameChecker.php]: Where is this list from?

File: src/PhpSpec/Listener/MethodNotFoundListener.php [hunk 4]
```diff
@@ -82,4 +106,19 @@ public function afterSuite(SuiteEvent $event)
...
+    {
+        if (!$this->nameChecker->supports($methodName)) {
+            $this->wrongMethodNames[] = $methodName;
+        }
+    }
+
+    private function writeErrorMessage()
+    {
+        foreach ($this->wrongMethodNames as $methodName) {
+            $message = sprintf("You cannot use restricted `%s` as a method name", $methodName);
```
Comment (ciaranmcnulty) [inline src/PhpSpec/Listener/MethodNotFoundListener.php]: Maybe instead of `restricted` use `the reserved word`

**Comment by ciaranmcnulty** [comment]
I agree with @avant1 about case issues,

I made some naming comments but the overall code is solid

File: src/PhpSpec/Util/PhpNameChecker.php (hunk 3 above)
Comment (mheki) [inline src/PhpSpec/Util/PhpNameChecker.php]: http://php.net/manual/en/reserved.keywords.php

**Comment by mheki** [comment]
done

File: src/PhpSpec/Listener/MethodNotFoundListener.php [hunk 5]
```diff
@@ -27,12 +29,33 @@ class MethodNotFoundListener implements EventSubscriberInterface
...
+        IO $io,
+        ResourceManagerInterface $resources,
+        GeneratorManager $generator,
+        NameCheckerInterface $nameChecker = null
+    ) {
         $this->io        = $io;
         $this->resources = $resources;
         $this->generator = $generator;
+
+        if (null === $nameChecker) {
```
Comment (ciaranmcnulty) [inline src/PhpSpec/Listener/MethodNotFoundListener.php]: In other places we've done this with a ternary:

```
$this->nameChecker = $nameChecker ? $nameChecher : new ReservedWordsMethodNameChecker();
```

(one day we can use the `?:` operator when we drop old PHP versions

File: src/PhpSpec/Listener/MethodNotFoundListener.php [hunk 6]
```diff
@@ -63,6 +88,11 @@ public function afterSuite(SuiteEvent $event)
             return;
         }
 
+        if (!empty($this->wrongMethodNames)) {
```
Comment (ciaranmcnulty) [inline src/PhpSpec/Listener/MethodNotFoundListener.php]: `empty` is redundant here, empty arrays are always `false` and if it stops being an array I'd prefer a hard break

**Comment by mheki** [comment]
I've just found (and pushed changes) that `CollaboratorMethodNotFoundListener` needs the same functionality, now I'm almost certain this bugfix should be done with additional event listener instead of repeating functionality all over again

File: src/PhpSpec/Console/IO.php [hunk 7]
```diff
@@ -342,4 +342,25 @@ public function getBlockWidth()
...
         return $width;
     }
+
+    /**
+     * @param string $message
+     * @param int|null $indent
+     */
+    public function writeError($message, $indent = null)
+    {
+        $message = wordwrap($message, $this->getBlockWidth() - 4, "\n", true);
```
Comment (stof) [inline src/PhpSpec/Console/IO.php]: why ``- 4`` here ?

File: src/PhpSpec/Listener/CollaboratorMethodNotFoundListener.php [hunk 8]
```diff
@@ -48,15 +50,31 @@ class CollaboratorMethodNotFoundListener implements EventSubscriberInterface
...
+    public function __construct(
+        IO $io,
+        ResourceManagerInterface $resources,
+        GeneratorManager $generator,
+        NameCheckerInterface $nameChecker = null
+    ) {
         $this->io = $io;
         $this->resources = $resources;
         $this->generator = $generator;
+        $this->nameChecker = $nameChecker ? $nameChecker : new ReservedWordsMethodNameChecker();
```
Comment (stof) [inline src/PhpSpec/Listener/CollaboratorMethodNotFoundListener.php]: this can use a shorthand notation

Comment (ciaranmcnulty) [inline src/PhpSpec/Listener/CollaboratorMethodNotFoundListener.php]: You're right, for some reason I thought `?:` was 5.4

File: src/PhpSpec/Console/IO.php (hunk 7 above)
Comment (mheki) [inline src/PhpSpec/Console/IO.php]: to be consistent with `askConfirmation()` https://github.com/phpspec/phpspec/blob/master/src/PhpSpec/Console/IO.php#L274

Comment (mheki) [inline src/PhpSpec/Console/IO.php]: if we want to keep it more generic, it should be `$indent * 2` tbh, then the error message is displayed with the same padding on both sides

Comment (ciaranmcnulty) [inline src/PhpSpec/Console/IO.php]: That sounds good.

**Comment by mheki** [comment]
I've changed wording and the block itself, so now corresponds to the `broken code` indications. When the method is wrong, runner will not run the suite again.
screen shot 2016-02-16 at 16 08 39

**Comment by ciaranmcnulty** [comment]
Great, thanks @mheki!

**Pull request merged**